restapi/src/test/java/io/stargate/db/datastore/ValidatingDataStore.java

```diff
@@ -96,22 +96,36 @@ public boolean supportsLoggedBatches() {
         .findAny()
         .orElseGet(() -> Assertions.fail("Unexpected query: " + queryString));
 
-    Prepared<B> p = new Prepared<>(query);
-    prepared.add(p);
-    return CompletableFuture.completedFuture(p);
+    // Generate unique IDs for each `prepare` call to validate that all prepared statements get
+    // executed.
+    ByteBuffer buffer = ByteBuffer.allocate(Integer.BYTES);
+    buffer.putInt(preparedIdSeq.incrementAndGet());
+    MD5Digest preparedId = MD5Digest.wrap(buffer.array());
+    Query<B> queryWithId = query.withPreparedId(preparedId);
+    prepared.computeIfAbsent(preparedId, p -> new Prepared<>(queryWithId));
+    return CompletableFuture.completedFuture(queryWithId);
   }
 
   @Override
   public CompletableFuture<ResultSet> execute(
       BoundQuery query, UnaryOperator<Parameters> parametersModifier) {
-    ExpectedExecution execution;
-    if (query instanceof ExpectedExecution) {
-      execution = (ExpectedExecution) query;
-    } else {
-      Prepared<?> prepared = (Prepared<?>) prepare(query.source().query()).join();
-      execution = (ExpectedExecution) prepared.bindValues(query.values());
+    Optional<MD5Digest> preparedId = query.source().query().preparedId();
+    if (!preparedId.isPresent()) {
+      preparedId = prepare(query.source().query()).join().preparedId();
```

Comment:
why do we use a blocking operation (`join`) in the method that is non-blocking (returns `CompletableFuture<ResultSet>`)? Shouldn't it be fixed?

Reply:
duplicate comment?... resolving